controllers/control/k8ssandratask_controller.go

```diff
@@ -0,0 +1,212 @@
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package control
+
+import (
+	"context"
+	"fmt"
+	cassapi "github.com/k8ssandra/cass-operator/apis/control/v1alpha1"
+	k8sapi "github.com/k8ssandra/k8ssandra-operator/apis/k8ssandra/v1alpha1"
+	"github.com/k8ssandra/k8ssandra-operator/pkg/clientcache"
+	"github.com/k8ssandra/k8ssandra-operator/pkg/config"
+	"github.com/k8ssandra/k8ssandra-operator/pkg/labels"
+	"github.com/k8ssandra/k8ssandra-operator/pkg/utils"
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/cluster"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+
+	"k8s.io/apimachinery/pkg/runtime"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+
+	api "github.com/k8ssandra/k8ssandra-operator/apis/control/v1alpha1"
+)
+
+// K8ssandraTaskReconciler reconciles a K8ssandraTask object
+type K8ssandraTaskReconciler struct {
+	*config.ReconcilerConfig
+	client.Client
+	Scheme      *runtime.Scheme
+	ClientCache *clientcache.ClientCache
+}
+
+//+kubebuilder:rbac:groups=control.k8ssandra.io,namespace="k8ssandra",resources=k8ssandratasks,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups=control.k8ssandra.io,namespace="k8ssandra",resources=k8ssandratasks/status,verbs=get;update;patch
+//+kubebuilder:rbac:groups=control.k8ssandra.io,namespace="k8ssandra",resources=k8ssandratasks/finalizers,verbs=update
+
+func (r *K8ssandraTaskReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	logger := log.FromContext(ctx).WithValues("K8ssandraTask", req.NamespacedName)
+
+	logger.Info("Fetching task", "K8ssandraTask", req.NamespacedName)
+	k8Task := &api.K8ssandraTask{}
+	err := r.Get(ctx, req.NamespacedName, k8Task)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			return ctrl.Result{}, nil
+		}
+		return ctrl.Result{RequeueAfter: r.ReconcilerConfig.DefaultDelay}, err
+	}
+	// TODO k8Task.DeepCopy()? why?
+
+	if k8Task.DeletionTimestamp != nil {
+		// TODO delete dependent objects?
+		return ctrl.Result{}, nil
+	}
+
+	kcKey := k8Task.GetClusterKey()
+	logger.Info("Fetching cluster", "K8ssandraCluster", kcKey)
+	kc := &k8sapi.K8ssandraCluster{}
+	err = r.Get(ctx, kcKey, kc)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	dcs, err := filterDcs(kc, k8Task.Spec.Datacenters)
+	if err != nil {
+		return ctrl.Result{}, err
```

Comment:
Do we wish to entertain user with the knowledge that their spec is incorrect?

Reply:
We should surface the error in the status in some way. Consider trying to build some automation around this such that we should get notification if the tasks fails or is unable to run. We would grep the operator logs, but that would be brittle. It would easier and better to be able to query the status.

Reply:
But we do not tell the user where that DC is wrong. Lets say in rebuild, was it incorrect in the source_datacenter or target datacenters?

Reply:
I've since switched to Kubernetes events for user error reporting (see the docs added by this PR).